### PR TITLE
Issue 9519 super init with non self arg

### PR DESCRIPTION
**Pull request by jzohrab**
Work-in-progress for issue #9519, to see if I can crack it.

This is going to contain many junk commits that I'll rebase and get rid of later, just doing this so that people can see it and maybe comment.

The first commit adds a failing test case, then there is some debug crap just so I can understand the code as I go.

### CODE REVIEW

**Comment by jzohrab** [comment]
the test file lines 11-13

```
class Win(Window):
    def __init__(self, txt):
        super().__init__(txt)   # the bad line
```

Running `pytest -k test_super_init_with_non_self_argument -vv`, line 13 stops debugging at point i

```
     -------------------------
     visit call, node: Call(func=<Attribute.__init__ l.13 at 0x105fd42d0>,
         args=[<Name.txt l.13 at 0x106131fd0>],
         keywords=[])
     a
     b
     c
     d
     f
     g
     i
```

which is this:

```
        _dp("f")
        if called.args.args is None:
            _dp("g")
            if called.name == "isinstance":
                # Verify whether second argument of isinstance is a valid type
                _dp("h")
                self._check_isinstance_args(node, callable_name)
            # Built-in functions have no argument information.
            _dp("i")
            return

        _dp("j")
```

so the function actually exits earlier than what was suggested in the original issue.

**Comment by jzohrab** [comment]
Current debug cruft:

```
     visit call, node: Call(func=<Attribute.__init__ l.13 at 0x108038190>,
         args=[<Name.txt l.13 at 0x1078c26d0>],
         keywords=[])
     Data dump for __init__ call
     node frame: FunctionDef.__init__(name='__init__',
                         position=Position(lineno=12, col_offset=4, end_lineno=12, end_col_offset=16),
                         decorators=None,
                         args=<Arguments l.12 at 0x107f7d9d0>,
                         returns=None,
                         type_params=[],
                         doc_node=None,
                         body=[<Expr l.13 at 0x107fbce60>])
     isinst: False
     funcdef: False
     called: <BoundMethod __init__ of builtins.object at 0x4429550992
     frame body: [<Expr l.13 at 0x107fbce60>]
     called in frame body: False
     npa: 1
     dec names: set()
     called.args.args is None
     called.name: __init__
     Returning now
```

**Comment by jzohrab** [comment]
Just prior to turning off the debug cruft, here is the output:

```
  -  visit call, node: Call(func=<Attribute.__init__ l.13 at 0x104b14190>,
  -      args=[<Name.txt l.13 at 0x10439e6d0>],
  -      keywords=[])
  -  Data dump for __init__ call
  -  call_site: <astroid.arguments.CallSite object at 0x104acc8a0>
  -  call site positional args:: [<Name.txt l.13 at 0x10439e6d0>]
  -  call site keyword args:: {}
  -  call site invalid args: False
  -  call site inv keywords: False
  -  node args: [<Name.txt l.13 at 0x10439e6d0>]
  -  node frame: FunctionDef.__init__(name='__init__',
  -                      position=Position(lineno=12, col_offset=4, end_lineno=12, end_col_offset=16),
  -                      decorators=None,
  -                      args=<Arguments l.12 at 0x104a599d0>,
  -                      returns=None,
  -                      type_params=[],
  -                      doc_node=None,
  -                      body=[<Expr l.13 at 0x104a98e60>])
  -  isinst: False
  -  funcdef: False
  -  called: <BoundMethod __init__ of builtins.object at 0x4373829008
  -  called.args: Arguments(vararg=None,
  -           kwarg=None,
  -           args=None,
  -           defaults=None,
  -           kwonlyargs=[],
  -           posonlyargs=[],
  -           posonlyargs_annotations=[],
  -           kw_defaults=None,
  -           annotations=[],
  -           varargannotation=None,
  -           kwargannotation=None,
  -           kwonlyargs_annotations=[],
  -           type_comment_args=[],
  -           type_comment_kwonlyargs=[],
  -           type_comment_posonlyargs=[])
  -  frame body: [<Expr l.13 at 0x104a98e60>]
  -  called in frame body: False
  -  npa: 1
  -  dec names: set()
    ************* Module super_init_with_non_self_argument
  + tests/regrtest_data/super_init_with_non_self_argument.py:11:0: blah
  - tests/regrtest_data/super_init_with_non_self_argument.py:13:8: E1121: Too many positional arguments for __init__ call (too-many-function-args)
  -  called.args.args is None
  -  called.name: __init__
  -  Returning now
  - 
```

**Comment by jzohrab** [comment]
There's a bunch of cruft in the commits and the code, but the idea is there, so I'm going to mark this as ready for review.

I'm sure the actual ast checking for the `__init__` call can be done better but I'm not sure how to do it; I came to the current state after putting in all of the debug cruft during my hacking with the test.  It's a start, anyway.

**Comment by jzohrab** [comment]
As you can probably tell from my hacking, I'm very new at ast hacking, so there may be a far better way to do it than what I have.  At least the test case should be good as a starting point.

**Comment by jacobtylerwalls** [comment]
Hey, thanks for tackling this one!

What do you think about something like this?

```diff
diff --git a/pylint/checkers/typecheck.py b/pylint/checkers/typecheck.py
index edef5188b..afed2d3fe 100644
--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -1453,8 +1453,19 @@ accessed. Python regular expressions are accepted.",
         """Check that called functions/methods are inferred to callable objects,
         and that passed arguments match the parameters in the inferred function.
         """
-        called = safe_infer(node.func, compare_constructors=True)
+        is_super = False
+        if (
+            isinstance(node.func, nodes.Attribute)
+            and node.func.attrname == "__init__"
+            and isinstance(node.func.expr, nodes.Call)
+            and isinstance(node.func.expr.func, nodes.Name)
+            and node.func.expr.func.name == "super"
+        ):
+            inferred = safe_infer(node.func.expr)
+            if isinstance(inferred, astroid.objects.Super):
+                is_super = True
 
+        called = safe_infer(node.func, compare_constructors=True)
         self._check_not_callable(node, called)
 
         try:
@@ -1468,7 +1479,21 @@ accessed. Python regular expressions are accepted.",
             if called.name == "isinstance":
                 # Verify whether second argument of isinstance is a valid type
                 self._check_isinstance_args(node, callable_name)
-            # Built-in functions have no argument information.
+            # Built-in functions have no argument information, but we know
+            # super() only takes self.
+            if is_super and utils.is_builtin_object(called):
+                if (
+                    (call_site := astroid.arguments.CallSite.from_call(node))
+                    and call_site.positional_arguments
+                    and (first_arg := call_site.positional_arguments[0])
+                    and not (isinstance(first_arg, nodes.Name) and first_arg.name == "self")
+                ):
+                    self.add_message(
+                        "too-many-function-args",
+                        node=node,
+                        args=(callable_name,),
+                    )
+
             return
 
         if len(called.argnames()) != len(set(called.argnames())):
```

**Comment by jzohrab** [comment]
Hi @jacobtylerwalls - thanks very much.  I confess that this is out of my range of experience and so I can't say that your suggestion is the right way to handle it!  It _looks_ right to the uninformed (aka me), but I'm flying blind.

If you'd like, I can make the change you suggested in the branch, remove the cruft code comments, and push (without squashing).

**Comment by jacobtylerwalls** [comment]
Hey @jzohrab that would be great. Once you get a clean test run (I think I remember seeing there were a couple functional tests that needed updating?) the CI will run a "primer" job showing results of making this change on lint runs on several python packages. That might help us see if my suggestions is missing any edge cases. All part of the fun!

(I'm not saying my sketch is perfect, but I was trying to rely less on asserting on the str()ingified value of a node and things like that.)

**Comment by jzohrab** [comment]
Rebased off of upstream/main, added the changes suggested by @jacobtylerwalls, moved the is_super check to a separate function for readability.  Will see where that leaves us at the end of the run.  I don't _think_ there will be any new test failures introduced with this change, but I'm not sure.  It's possible that there are some existing test failures (which should be deactivated/indicated in a separate PR, maybe).

**Comment by jzohrab** [comment]
There are failures:

```
=================================== FAILURES ===================================
____________________ test_functional[metaclass_attr_access] ____________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7f0445c2cd00>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "metaclass_attr_access.py":
E       
E       Unexpected in testdata:
E          8: too-many-function-args (times 1)

pylint/testutils/lint_module_test.py:[142](https://github.com/pylint-dev/pylint/actions/runs/13796483619/job/38589265264?pr=10190#step:7:143): AssertionError
________________________ test_functional[super_checks] _________________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7f043aaf4600>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "super_checks.py":
E       
E       Unexpected in testdata:
E         61: too-many-function-args (times 1)
```

but I have no idea how to approach them!  Here's the metaclass_attr_access file:

```
class Meta(type):
    """the meta class"""
    def __init__(cls, name, bases, dictionary):
        super(Meta, cls).__init__(name, bases, dictionary)
        print(cls, cls._meta_args)
        delattr(cls, '_meta_args')


class Test:
    """metaclassed class"""
    __metaclass__ = Meta
    _meta_args = ('foo', 'bar')

    def __init__(self):
        print('__init__', self)
```

so maybe the super code suggested is handling super(Meta, cls) incorrectly.  Not sure, and I think I might be breaking things if I hack at it further.  Maybe someone with more experience with asts etc can take my work as a startisg point.

**Comment by jacobtylerwalls** [comment]
Hey thanks for the push. What I can say at a glance is that the pylint annotations you see in the files tab of this PR show some false positives with my suggestion -- looks like the logic needs to be fixed to make sure that the inference result for the class being init'ed is really `object` and not something else that might indeed take non-self args. Does that help?
